### PR TITLE
[SPARK-34414][SQL] OptimizeMetadataOnlyQuery should only apply for deterministic filters

### PR DESCRIPTION
**Pull request by yeshengm**
### What changes were proposed in this pull request?

Add a predicate to guard against non-deterministic filters on partitioned table to be optimized by the optimization rule OptimizeMetadataOnlyQuery.

### Why are the changes needed?

Similar to FileSourcePartitionPruning, OptimizeMetadataOnlyQuery should only apply for deterministic filters. If filters are non-deterministic, they have to be evaluated against partitions separately.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing UTs.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@yeshengm do you mind filing a JIRA please?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala [hunk 1]
```diff
@@ -52,8 +52,10 @@ case class OptimizeMetadataOnlyQuery(catalog: SessionCatalog) extends Rule[Logic
     plan.transform {
       case a @ Aggregate(_, aggExprs, child @ PhysicalOperation(
           projectList, filters, PartitionedRelation(partAttrs, rel))) =>
-        // We only apply this optimization when only partitioned attributes are scanned.
-        if (AttributeSet((projectList ++ filters).flatMap(_.references)).subsetOf(partAttrs)) {
+        // We only apply this optimization when only partitioned attributes are scanned and filters
+        // are deterministic.
+        if (AttributeSet((projectList ++ filters).flatMap(_.references)).subsetOf(partAttrs)
+          && filters.map(_.deterministic).forall(identity)) {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala:58]: Please add a test case

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala:58]: +1 for @gatorsmile 's advice for adding a test case.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala [hunk 2]
```diff
@@ -52,8 +52,10 @@ case class OptimizeMetadataOnlyQuery(catalog: SessionCatalog) extends Rule[Logic
     plan.transform {
       case a @ Aggregate(_, aggExprs, child @ PhysicalOperation(
           projectList, filters, PartitionedRelation(partAttrs, rel))) =>
-        // We only apply this optimization when only partitioned attributes are scanned.
-        if (AttributeSet((projectList ++ filters).flatMap(_.references)).subsetOf(partAttrs)) {
+        // We only apply this optimization when only partitioned attributes are scanned and filters
+        // are deterministic.
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala:56]: Thank you for updating comments. Could you revise the description (line 37) of this optimizer together.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala (hunk 1 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala:58]: indentation at line 58? We need two more leading spaces.

Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala:58]: nit: why not just `filters.forall(_.deterministic)`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala [hunk 3]
```diff
@@ -52,8 +52,10 @@ case class OptimizeMetadataOnlyQuery(catalog: SessionCatalog) extends Rule[Logic
     plan.transform {
       case a @ Aggregate(_, aggExprs, child @ PhysicalOperation(
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuery.scala:53]: I think `PhysicalOperation` only collects deterministic filters, isn't?

**Comment by c21** [comment]
After a second check, @yeshengm - can you help double check this is a valid fix? It seems that `PhysicalOperation` only returns deterministic filters here - https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala#L84-L92 .

**Pull request closed without merging**